### DISCUSSION TITLE
Circular entity mapping for discriminator map causing an error?

### DISCUSSION THREAD

**Discussion by DavidGarciaCat** [mikro-orm/mikro-orm#3680] [Q&A] [answered]
Good day,

I have one abstract class/entity with some child entities, mapping them with a discriminator map:

```ts
// src/entity/abstract-content.ts

@Entity({
    discriminatorColumn: 'content_type',
    discriminatorMap: { 'post': 'PostContent', 'pack': 'PackContent', 'images': 'ImagesContent' },
})
export default abstract class AbstractContent {
```

One of the child entities has an additional property mapping the abstract parent into a collection:

```ts
// src/entity/pack-content.ts

@Entity()
export default class PackContent extends AbstractContent {
    @OneToMany({ entity: () => AbstractContent, mappedBy: 'packContent' })
    public packContentItems: Collection<AbstractContent> = new Collection<AbstractContent>(this);
```

For the abstract entity, when I try to map one of the child items, I get an error. However, when I remove the mapping decorator, everything seems to work as expected.

```ts
// src/entity/abstract-content.ts

// ...
export default abstract class AbstractContent {
    // ...
    @ManyToOne({ entity: () => PackContent, inversedBy: 'packContentItems', nullable: true })
    public packContent: PackContent | undefined;
    // ...
```

If I comment this `@ManyToOne` decorator, the ORM init works as expected

If I keep this `@ManyToOne` decorator, then I get the error below
- `src/entity/pack-content.ts:5:49` is the `export default class PackContent extends AbstractContent` line
- `src/entity/abstract-content.ts:1:8` is the `import PackContent from "./pack-content.ts"` line

```json
{
    "level":"error",
    "message":"uncaughtException: Class extends value undefined is not a constructor or null",
    "trace":[
        {
            "column":49,
            "file":".../src/entity/pack-content.ts",
            "function":null,
            "line":5,
            "method":null,
            "native":false
        },

        // ... ( node_modules calls are listed here )

        {
            "column":1,
            "file":".../src/entity/abstract-content.ts",
            "function":null,
            "line":8,
            "method":null,
            "native":false
        }
    ]
}
```

Is there any limitation or restriction that might cause a circular reference, leading to an infinite loop, and it blocks the init process?

Maybe I need to set the `entity: () => AbstractContent` value to something different, where I can set child entities instead of the abstract class? If so, can I get an example, please?

I have another 10 entities mapped to this abstract class, and they are not throwing errors - just this one, which is a reference to a child class extending from the abstract one.

Thank you in advance, in case any help or guidance might be provided here.

**Comment by B4nan** [comment, marked as answer]
Afaik that error is not caused by the ORM, this is what you get if you try to extend `undefined` - which is what happens if your imports end up in a cycle. `AbstractContent` imports `PackContent` so it can use it in the property definition, while `PackContent` imports `AbstractContent` so it can extend it. Your import for `AbstractContent` in there ends up as `undefined` during the module resolution, which fails to be used for `extends` keyword.

**Reply by DavidGarciaCat** [reply]
I see... in that case, what might be better here? Maybe set `null` by default instead of `undefined`?

If so, I guess I might need to consider another approach for the getters and setters, as the ORM init config now has the `forceUndefined: true` parameter, so I will need to make sure I have the expected value to return the expected output when calling it

**Reply by DavidGarciaCat** [reply]
Actually, I can see the same error if I initialise the property as `null` so I guess this won't work

**Reply by B4nan** [reply]
this is not about any values on the class instance, this is about the class extension itself. you have a circular import, you need to get rid of it, otherwise your JS files cannot be interpretted.

ORM-wise you could use strings instead of class references to get around import cycles. maybe if you use just type import, things will work.

**Reply by DavidGarciaCat** [reply]
👍 I will give it a try right away - thanks for the tip

(and sorry if it sounds like a silly question, as I am still trying to get more familiar with Node/TS... I guess we all have been newbies at some point 😄)
